sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,652 @@ These follow the AAELF64_ base definition from Arm.
+second parameter is a structure ``__ifunc_arg_t`` defined in
+``sys/ifunc.h``, it provides access to all of the flags in HWCAP_ via
+fields of the structure.
+
+The IFUNC resolver function returns the address of the function
+implementation.
+
+IFUNC resolvers may be run when the dynamic linker is resolving relocations.
+additional restrictions on what they can contain.
+
+ * IFUNC resolvers must not be compiled with security features like
+   stack-protection, which requires a guard variable to be
+   initialized. Or instrumentation like ASAN that requires a shadow
+   map to be set up.
+
+ * IFUNC resolvers must not have a symbol binding of ``STB_WEAK``.
+
+The order of dynamic relocation resolution across an executable and
+all its shared libraries is platform specific. The following
+recommendations for writing IFUNC resolvers apply to the GNU glibc
+dynamic loader. Other dynamic linkers may have fewer requirements.
+
+ * An IFUNC resolver function must not call a function that may itself
+   require IFUNC initialization. If the IFUNC initialization for the
+   called function has not occured then undefined behavior results.
+
+ * In position-independent code an IFUNC resolver functions must not
+   call a function that requires a PLT entry. If the IFUNC resolver
+   runs as a result of a relocation in ``.rela.dyn`` then the
+   relocations in ``.rela.plt`` will not have been resolved. This means
+   that addresses in the ``.got.plt`` will be unchanged from their
+   static link time value.
+
+ * The IFUNC resolver function for a given function must be defined in
+   the same translation unit as the implementations of the function.
+
+ * IFUNC resolver functions must be idempotent. There can be
+   relocations in both ``.rela.dyn`` and ``.rela.plt`` to the same
+   IFUNC resolver function.
+
+IFUNC requirements for static linkers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Relocations to pre-emptable symbols of type ``STT_GNU_IFUNC`` are
+handled in the same way as symbols of type ``STT_FUNC``. The symbol
+type ``STT_GNU_IFUNC`` is propagated into the dynamic symbol table.
+
+Relocations to non-pre-empatable symbols of type ``STT_GNU_IFUNC`` are
+resolved to a PLT entry that loads the value of:
+
+ * A ``.got.plt`` entry for a non-branch relocation such as
+   ``R_AARCH64_ABS64``. The entry has an ``R_AARCH64_IRELATIVE``
+   dynamic relocation in ``.rela.dyn``.
+
+ * A ``.got.plt`` entry for a branch relocation such as
+   ``R_AARCH64_CALL``. A ``R_AARCH64_IRELATIVE`` dynamic relocation is
+   added to ``.rela.dyn`` or ``rela.plt``.
+
+Due to the ordering requirements on IFUNC resolvers the PLT entry and
+associated ``.got.plt`` entry are often implemented in a separate
+``.iplt`` and ``.iplt.got`` sections so that they are placed after the
+``.plt`` and ``.got.plt`` sections respectively.
+
+Like ``R_AARCH64_RELATIVE`` the ``R_AARCH64_IRELATIVE`` relocation
+does not require a symbol. For ``RELA`` type relocations the addend of
```

Comment:
Q: Does `does not require` indicate the intent that the symbol index must be zero?